File: pasku/BowlingGame.php
<?php
/**
 * Kata 2011-04: Bowling Game
 *
 * Kata trying to resolve the BowlingGame problem using TDD.
 *
 * @category   Kata
 * @author     Guillermo Pascual <@pasku1>
 * @version    0.1
 * @link       http://github.com/12meses12katas/Abril-Bowling
 */

class BowlingGame
{
    protected $rolls = array();

    public function roll($pins)
    {
        $this->rolls[] = $pins;
    }

    public function score()
    {
        $score      = 0;
        $frameIndex = 0;

        for ($frame = 0; $frame < 10; $frame++) {
            if ($this->isStrike($frameIndex)) {
                $score += 10 + $this->strikeBonus($frameIndex);
                $frameIndex++;
            }

            else if ($this->isSpare($frameIndex)) {
                $score += 10 + $this->spareBonus($frameIndex);
                $frameIndex += 2;
            }

            else {
                $score += $this->sumOfPinsInFrame($frameIndex);
                $frameIndex += 2;
            }
        }
        return $score;
    }

    protected function isStrike($frameIndex)
    {
        return $this->rolls[$frameIndex] === 10;
    }

    protected function isSpare($frameIndex)
    {
        return $this->sumOfPinsInFrame($frameIndex) === 10;
    }

    protected function sumOfPinsInFrame($frameIndex)
    {
        return $this->rolls[$frameIndex] +
               $this->rolls[$frameIndex + 1];
    }

    protected function strikeBonus($frameIndex)
    {
        return $this->rolls[$frameIndex + 1] +
               $this->rolls[$frameIndex + 2];
    }

    protected function spareBonus($frameIndex)
    {
        return $this->rolls[$frameIndex + 2];
    }
}

File: ggalmazor/java/src/kata/BowlingGame.java
package kata;

import java.util.ArrayList;
import java.util.List;

public class BowlingGame {

    private List<Integer> rolls = new ArrayList<Integer>();

    public void roll(int pins) {
        rolls.add(pins);
    }

    public Integer score() {
        int score = 0;
        int roll = 0;
        for (int frame = 1; frame <= 10; frame++) {
            if (isStrike(roll)) {
                score += 10 + strikeBonus(roll);
                roll += 1;
            } else if (isSpare(roll)) {
                score += 10 + spareBonus(roll);
                roll += 2;
            } else {
                score += normalFrameScore(roll);
                roll += 2;
            }
        }
        return score;
    }

    private boolean isStrike(int roll) {
        return 10 == rolls.get(roll);
    }

    private boolean isSpare(int roll) {
        return 10 == normalFrameScore(roll);
    }

    private int strikeBonus(int roll) {
        return normalFrameScore(roll + 1);
    }

    private Integer spareBonus(int roll) {
        return rolls.get(roll + 2);
    }

    private int normalFrameScore(int roll) {
        return rolls.get(roll) + rolls.get(roll + 1);
    }
}
